cmd/multi-cluster/cmd/api.go

```diff
@@ -0,0 +1,113 @@
+const (
+	UserCtxKey = iota
+	ClustersClientsPoolCtxKey
+)
+
+type User struct {
+	Email string
+	Team  string
+}
+
+// simulates OIDC autenticate that potentially adds the user to context
+func authMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Injecting authenticated user
+		ctx := context.WithValue(r.Context(), UserCtxKey, User{Email: "luiz.filho@weave.gitops", Team: "developers"})
+
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+// creates clusters client for provided user in the context
+func clustersClientsMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Making sure user is present in the request and has been authenticated
+		user, ok := r.Context().Value(UserCtxKey).(User)
+
+		if !ok && user.Email == "" && user.Team == "" {
+			w.WriteHeader(http.StatusUnauthorized)
+			fmt.Fprintln(w, `{"error": "there is no authorized user"}`) // TODO: use proper structs for this error message
+			return
+		}
+
+		// TODO: we should probably cache everything down here after we check the user is present,
+		// to avoid creating a bunch of clients in every single call
+		clustersFetcher, err := NewConfigMapClustersFetcher()
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintf(w, `{"error": "failed getting management cluster client: %s"}`, err) // TODO: use proper structs for this error message
+			return
+		}
+
+		clusters, err := clustersFetcher.Fetch(r.Context()) // TODO: cache it
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintln(w, `{"error": "failed fetching clusters list: %w"}`, err) // TODO: use proper structs for this error message
+			return
+		}
+
+		// TODO: cache it
+		clientsPool := NewClustersClientsPool()
+		for _, c := range clusters {
+			if err := clientsPool.Add(user, c); err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				fmt.Fprintf(w, `{"error": "failed adding cluster client to the pool: %s"}`, err) // TODO: use proper structs for this error message
+				return
+			}
+		}
+
+		ctx := context.WithValue(r.Context(), ClustersClientsPoolCtxKey, clientsPool)
+
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+func final(w http.ResponseWriter, r *http.Request) {
+	clientsPool, ok := r.Context().Value(ClustersClientsPoolCtxKey).(*clientsPool)
+	if !ok {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintln(w, `{"error": "failed fetching clusters clients"}`) // TODO: use proper structs for this error message
+
+		return
+	}
+
+	for name, c := range clientsPool.Clients() {
+		podList := &corev1.PodList{}
+
+		err := c.List(context.Background(), podList, client.InNamespace("developers"))
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintf(w, `{"error": "failed to list pods for cluster=%s"}`, name) // TODO: use proper structs for this error message
+
+			return
```

Comment:
Is it possible a given user has no access to all clusters, only a set of clusters? In this case, return would terminate the whole pool at the first cluster they have no access. At least as I see the list of clusters are global in the secret.

I imagine something like this:
1. There 6 clusters: dev1, dev2, staging1, staging2, prod1, prod2
2. Given user has access only to dev1, staging1, prod1
3. Another user has access to dev1, dev2, staging1, staging2

Reply:
I don't know where would it break

1. Here
2. When we try to add (`clientsPool.Add(user, c)`)
3. When we try to get the secret

Reply:
Regarding the access, it would break here. 

The actual connection to the leaf cluster is made using the `impersonator` secret, so it will always work unless there is a misconfiguration or the hub cluster can't reach the leaf cluster given a network issue.

But we should handle it anyway, and avoid returning immediately if a single cluster connection fails.